crypto/crypto_aes.c

```diff
@@ -63,30 +63,47 @@ static struct aes_test {
 
 /* Test hardware intrinsics against test vectors. */
 static int
-hwtest(struct aes_test * knowngood)
+hwtest(struct aes_test * knowngood,
+    int(* func)(const uint8_t *, size_t, const uint8_t[16], uint8_t[16]))
 {
-	void * kexp_hw;
-	uint8_t ctext_hw[16];
+	uint8_t ctext[16];
 
 	/* Sanity-check. */
 	assert((knowngood->len == 16) || (knowngood->len == 32));
 
-	/* Expand the key and encrypt with hardware intrinsics. */
+	/* Expand the key and encrypt with the provided function. */
+	if (func(knowngood->key, knowngood->len, knowngood->ptext, ctext))
+		goto err0;
+
+	/* Do the outputs match? */
+	return (memcmp(knowngood->ctext, ctext, 16));
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
 #if defined(CPUSUPPORT_X86_AESNI)
-	if ((kexp_hw = crypto_aes_key_expand_aesni(knowngood->key,
-	    knowngood->len)) == NULL)
+static int
+test_x86_aesni(const uint8_t * key, size_t len, const uint8_t ptext[16],
```

Comment:
It occurred to me that we could name this function `x86_aesni_oneshot()` or `x86_aesni_Buf()`, since this function itself isn't "testing" anything.

(The name `x86_aesni_Buf()` follows the pattern of `sha256.h` such as `SHA256_Buf()`, but I hate the capitalization.  I only mention that name as a "brainstorming" thing.)